src/java.base/unix/native/libnio/ch/FileChannelImpl.c

```diff
@@ -248,6 +259,38 @@ Java_sun_nio_ch_FileChannelImpl_transferTo0(JNIEnv *env, jobject this,
 #endif
 }
 
+JNIEXPORT jlong JNICALL
+Java_sun_nio_ch_FileChannelImpl_transferFrom0(JNIEnv *env, jobject this,
+                                              jobject srcFDO, jobject dstFDO,
+                                              jlong position, jlong count)
+{
+#if defined(__linux__)
+    if (my_copy_file_range_func == NULL)
+        return IOS_UNSUPPORTED;
+
+    jint srcFD = fdval(env, srcFDO);
+    jint dstFD = fdval(env, dstFDO);
+
+    off64_t offset = (off64_t)position;
+    jlong n = my_copy_file_range_func(srcFD, NULL, dstFD, &offset, count, 0);
+    if (n < 0) {
+        if (errno == EAGAIN)
+            return IOS_UNAVAILABLE;
+        if ((errno == EBADF || errno == EINVAL || errno == EXDEV) &&
+            ((ssize_t)count >= 0))
+            return IOS_UNSUPPORTED_CASE;
```

Comment:
I don't think you can return IOS_UNSUPPORTED_CASE for the these cases because it will disable the use of copy_file_range for the everyone and all channels. So I think we'll need to think through whether the append and overlap cases where we checked first or return IOS_UNSUPPORTED_CASE does not disable it for everyone. I think the former is probably better.

Reply:
Isn't it that returning `IOS_UNSUPPORTED` would disable it for everyone? I thought that `IOS_UNSUPPORTED_CASE` allows it to be tried again. Of course one could check for append or overlap in the Java layer instead. I actually had it that way initially for the overlap case but removed it.

Reply:
> Isn't it that returning `IOS_UNSUPPORTED` would disable it for everyone? I thought that `IOS_UNSUPPORTED_CASE` allows it to be tried again. Of course one could check for append or overlap in the Java layer instead. I actually had it that way initially for the overlap case but removed it.

So I see, it's only IOS_UNSUPPORTED that disables it. With IOS_UNSUPPORTED_CASE then you may retry it again.

Reply:
Correct. It is IOS_UNSUPPORTED for example if the copy_file_range function pointer is NULL (not available via dlsym()). Then of course one can do nothing.

Reply:
Okay, let's go with this.